starknet_py/utils/typed_data/typed_data.py

```diff
@@ -35,16 +33,25 @@ class TypedData:
 
     @staticmethod
     def from_dict(data: TypedDataDict) -> "TypedData":
-        return TypedDataSchema().load(data)
+        return cast(TypedData, TypedDataSchema().load(data))
+
+    def _encode_value(self, type_name: str, value: Union[int, str, dict, list]) -> int:
+        if is_pointer(type_name) and isinstance(value, list):
+            type_name = strip_pointer(type_name)
+
+            # if type_name is not a struct
+            if type_name not in self.types:
+                return compute_hash_on_elements(
+                    [int(get_hex(val), 16) for val in value]
+                )
 
-    def _encode_value(self, type_name: str, value: Union[int, str]) -> int:
-        if is_pointer(type_name):
             return compute_hash_on_elements(
-                [self.struct_hash(strip_pointer(type_name), data) for data in value]
+                [self.struct_hash(type_name, data) for data in value]
             )
-        if type_name in self.types:
+        # if type_name is a struct
+        if type_name in self.types and isinstance(value, dict):
             return self.struct_hash(type_name, value)
-        return int(get_hex(value), 16)
+        return int(get_hex(str(value)), 16)
```

Comment:
pyright; without the cast:

Argument of type "int | str | dict[Unknown, Unknown] | list[Unknown]" cannot be assigned to parameter "value" of type "int | str" in function "get_hex"

Reply:
Okay, will `value` always be castable to hash?

Reply:
`def get_hex(value: Union[int, str]) -> str:`

Reply:
Perhaps a safer approach here would be to `value = typing.cast(Union[int, str], value)` instead?

Reply:
If so, should these `isinstance` checks be changed to casting as well?
```
if type_name in self.types and isinstance(value, dict):
    return self.struct_hash(type_name, value)
```
```
if is_pointer(type_name) and isinstance(value, list):
```
(since they are redundant, i.e if the first part of the condition is `True` then the instance check will return `True` as well)

Reply:
They can be changed, though it doesn't matter that much since they don't mutate the data like `str` does.